u_parrots.go

```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
 	}
 	return s[:sliceLen]
 }
+
+func FingerprintClientHello(data []byte) (*ClientHelloSpec, error) {
+	m := &clientHelloMsg{raw: data}
+
+	clientHelloSpec := &ClientHelloSpec{}
+
+	parse := func() bool {
+		// Copied directly from tls/handshake_messages.go clientHelloMsg.unmarshal
+		s := cryptobyte.String(data)
+		if !s.Skip(4) || // message type and uint24 length field
+			!s.ReadUint16(&m.vers) || !s.ReadBytes(&m.random, 32) ||
+			!readUint8LengthPrefixed(&s, &m.sessionId) {
+			return false
+		}
+
+		var cipherSuites cryptobyte.String
+		if !s.ReadUint16LengthPrefixed(&cipherSuites) {
+			return false
+		}
+		m.cipherSuites = []uint16{}
+		m.secureRenegotiationSupported = false
+		for !cipherSuites.Empty() {
+			var suite uint16
+			if !cipherSuites.ReadUint16(&suite) {
+				return false
+			}
+			if suite == scsvRenegotiation {
+				m.secureRenegotiationSupported = true
+			}
+			m.cipherSuites = append(m.cipherSuites, suite)
+		}
+
+		if !readUint8LengthPrefixed(&s, &m.compressionMethods) {
+			return false
+		}
+
+		// utls specific
+		clientHelloSpec.CipherSuites = m.cipherSuites
```

Comment:
The cipher suites may contain a GREASE value, which in many cases be expected to change between hellos.  In the parrots, the location of the grease value is replaced by GREASE_PLACEHOLDER, and a new GREASE value is selected when generating a new hello.  Same note as elsewhere, there is only one selection method available (the random selection of boringssl)... may be safe to assume.

Reply:
oic, you've adjusted the generation to select a new grease value when it's GREASE_PLACEHOLDER or any valid grease value?

Reply:
Yeah inside `ApplyPreset` (the function that takes the `ClientHelloSpec` and populates the hello struct), it's already doing the regreasing. I just had to change the check:
https://github.com/getlantern/utls/pull/6/files#diff-17a6357df708d056c23f2eb7f0c63069R615-R661